src/assign.c

```diff
@@ -381,12 +381,13 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
     for (int i=0; i<targetlen; ++i) {
       if ((rowsd[i]<0 && rowsd[i]!=NA_INTEGER) || rowsd[i]>nrow)
         error(_("i[%d] is %d which is out of range [1,nrow=%d]"), i+1, rowsd[i], nrow);  // set() reaches here (test 2005.2); := reaches the same error in subset.c first
-      if (rowsd[i]>=1) numToDo++;
+      if (rowsd[i]>=0) numToDo++;
     }
     if (verbose) Rprintf(_("Assigning to %d row subset of %d rows\n"), numToDo, nrow);
     // TODO: include in message if any rows are assigned several times (e.g. by=.EACHI with dups in i)
     if (numToDo==0) {
-      if (!length(newcolnames)) {
+      // isString(cols) is exclusive to calls from set(), not :=, this part should only bind for set()
```

Comment:
not sure I'm understanding:

 - `isString()` is for `set()`
 - this branch doesn't apply when `!isString()`
 - The comment says "this part should only bind for `set()`"

Maybe I'm just not understanding what we mean by "this part should bind". We could possible just exclude everything after `:=`, that may be good enough.